omod/src/main/java/org/openmrs/web/controller/ForgotPasswordFormController.java

```diff
@@ -143,45 +149,78 @@ protected ModelAndView onSubmit(HttpServletRequest request, HttpServletResponse
-					for (int i = 0; i < 8; i++) {
-						randomPassword.append(String.valueOf((Math.random() * (127 - 48) + 48)));
-					}
-					
-					try {
-						Context.addProxyPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS);
-						//Context.getUserService().changePassword(user, randomPassword.toString());
-					}
-					finally {
-						Context.removeProxyPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS);
-					}
-					
-					httpSession.setAttribute("resetPassword", randomPassword);
-					httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.password.reset");
-					Context.authenticate(username, randomPassword.toString());
-					httpSession.setAttribute("loginAttempts", 0);
-					return new ModelAndView(new RedirectView(request.getContextPath() + "/options.form#Change Login Info"));
-				} else {
+				if (user == null) {
 					httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.answer.invalid");
 					httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.question.fill");
-					//request.setAttribute("secretQuestion", user.getSecretQuestion());
+					request.setAttribute("secretQuestion", getRandomFakeSecretQuestion(username));
+				} else {
+					String secretQuestion = Context.getUserService().getSecretQuestion(user);
+					if (secretQuestion == null || secretQuestion.equals("")) {
+						httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.question.empty");
+					} else if (secretQuestion != null && Context.getUserService().isSecretAnswer(user, secretAnswer)) {
+						
+						String randomPassword = getRandomPassword();
+						
+						try {
+							Context.addProxyPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS);
+							Context.getUserService().changePassword(user, randomPassword);
+						} finally {
+							Context.removeProxyPrivilege(PrivilegeConstants.EDIT_USER_PASSWORDS);
+						}
+						
+						httpSession.setAttribute("resetPassword", randomPassword);
+						httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.password.reset");
+						Context.authenticate(username, randomPassword);
+						httpSession.setAttribute("loginAttempts", 0);
+						
+						return new ModelAndView(new RedirectView(request.getContextPath() + "/options.form#Change Login Info"));
+					} else {
+						httpSession.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, "auth.answer.invalid");
+						httpSession.setAttribute(WebConstants.OPENMRS_MSG_ATTR, "auth.question.fill");
+						request.setAttribute("secretQuestion", secretQuestion);
+					}
 				}
 			}
 		}
 		
 		loginAttemptsByIP.put(ipAddress, forgotPasswordAttempts);
+		
 		request.setAttribute("uname", username);
+		
 		return showForm(request, response, errors);
 	}
 	
-}
+	public String getRandomPassword() {
+		//Password should be of length at-least 8, must have 1 upper case letter and 1 number
+		return RandomStringUtils.randomAlphabetic(1).toUpperCase() + RandomStringUtils.randomAlphanumeric(8)+ RandomStringUtils.randomNumeric(1);
```

Comment:
The generated password needs to be longer than the value of the required minimum password length, an arbitrary number like 8 should be used as the default value  in case the global property is not set otherwise this will fail if the amdin has set the global property to a longer value like 12